README.md
# MVTImageryProvider

Render Mapbox style in CesiumJs. This project is very simple, because the complex rendering task is compeleted by mapbox-gl-js, you should also check [Mapbox-vector-tiles-basic-js-renderer](https://github.com/landtechnologies/Mapbox-vector-tiles-basic-js-renderer) for more detail.

![mvt-basic-render](http://img.badgesize.io/https://unpkg.com/mvt-basic-render?compression=gzip&label=gzip) ![mvt-imagery-provider](http://img.badgesize.io/https://unpkg.com/mvt-imagery-provider?compression=gzip&label=gzip) <a href="https://www.npmjs.com/package/mvt-imagery-provider">![](https://img.shields.io/npm/v/mvt-imagery-provider)</a>

## Install

```bash
#npm
npm install --save mvt-imagery-provider
#pnpm
pnpm add mvt-imagery-provider
```

## Usage

```ts
import * as Cesium from "cesium";
import MVTImageryProvider from 'mvt-imagery-provider';

const cesiumViewer = new Cesium.Viewer("cesiumContainer");

const provider = new MVTImageryProvider({
  style: 'https://demotiles.maplibre.org/style.json',
  accessToken: MAPBOX_TOKEN
});

cesiumViewer.imageryLayers.addImageryProvider(provider);

```

## API

```ts
class MVTImageryProvider {
  /**
   * create a MVTImageryProvider Object
   * @param {MVTImageryProviderOptions} options MVTImageryProvider options as follow:
   * @param {Resource | string | StyleSpecification} options.style - mapbox style object or url Resource.
   * @param {string} options.accessToken - mapbox style accessToken.
   * @param {RequestTransformFunction} options.transformRequest - use transformRequest to modify tile requests.
   * @param {Number} [options.tileSize = 256] - can be 256 or 512. defaults to 256.
   * @param {Number} [options.maximumLevel = 18] - if cesium zoom level exceeds maximumLevel, layer will be invisible, defaults to 18.
   * @param {Number} [options.minimumLevel = 0] - if cesium zoom level belows minimumLevel, layer will be invisible, defaults to 0.
   * @param {Boolean} [options.showCanvas = false] - if show canvas for debug.
   * @param {Boolean} [options.enablePickFeatures = true] - enable pickFeatures or not, defaults to true.
   * @param {Function} options.sourceFilter - sourceFilter is used to filter which source participate in pickFeature process.
   * @param {WebMercatorTilingScheme | GeographicTilingScheme} [options.tilingScheme = WebMercatorTilingScheme] - Cesium tilingScheme, defaults to WebMercatorTilingScheme(EPSG: 3857).
   * @param {Credit} options.credit - A credit contains data pertaining to how to display attributions/credits for certain content on the screen.
   * @example
   * const imageryProvider = new MVTImageryProvider({
        style: 'https://demotiles.maplibre.org/style.json'
      });
  */
  constructor(options: MVTImageryProviderOptions);
  /**
   * get mapbox style json obj
   */
  get style(): StyleSpecification;
  get isDestroyed(): boolean;
  requestImage(x: number, y: number, level: number, releaseTile?: boolean): Promise<HTMLImageElement | HTMLCanvasElement | any> | undefined;
  pickFeatures(x: number, y: number, zoom: number, longitude: number, latitude: number): Promise<ImageryLayerFeatureInfo[]> | undefined;
  destroy(): void;
}

type MVTImageryProviderOptions = {
  style: string | Resource | StyleSpecification;
  /** accessToken needed if mapbox style not public */
  accessToken?: string;
  /** forceHTTPS defaults false */
  /**
   * A `RequestParameters` object to be returned from Map.options.transformRequest callbacks.
   * @return {Object} RequestParameters
   * @property {string} url The URL to be requested.
   * @property {Object} headers The headers to be sent with the request.
   * @property {string} method Request method `'GET' | 'POST' | 'PUT'`.
   * @property {string} body Request body.
   * @property {string} type Response body type to be returned `'string' | 'json' | 'arrayBuffer'`.
   * @property {string} credentials `'same-origin'|'include'` Use 'include' to send cookies with cross-origin requests.
   * @property {boolean} collectResourceTiming If true, Resource Timing API information will be collected for these transformed requests and returned in a resourceTiming property of relevant data events.
   * @example
   * // use transformRequest to modify requests that begin with `http://myHost`
   * transformRequest: function(url, resourceType) {
   *  if (resourceType === 'Source' && url.indexOf('http://myHost') > -1) {
   *    return {
   *      url: url.replace('http', 'https'),
   *      headers: { 'my-custom-header': true },
   *      credentials: 'include'  // Include cookies for cross-origin requests
   *    }
   *   }
   *  }
   */
  transformRequest?: RequestTransformFunction;
  showCanvas?: boolean;
  tileSize?: number;
  maximumLevel?: number;
  minimumLevel?: number;
  credit?: string;
  hasAlphaChannel?: boolean;
  sourceFilter?: any;
  tilingScheme?: WebMercatorTilingScheme | GeographicTilingScheme;
};
```

## StyleSpecification

Reference to the [Mapbox Style Specification](https://docs.mapbox.com/mapbox-gl-js/style-spec/root/), the mapbox-gl version is 0.43.0

## Demo

[online Demo](https://mvti-magery-provider.vercel.app/)

Launch the app in the demo folder, and then visit http://localhost:8080/

```node
pnpm install
npm start
```

| ![bguCSe.png](https://s1.ax1x.com/2022/07/28/vCk4z9.png) | ![macrostrat.png](https://s1.ax1x.com/2022/07/28/vCkcZV.png) |
| ------- | ------- |

## Credit

https://github.com/kikitte/MVTImageryProvider


CHANGELOG.md
## [0.7.2](https://github.com/hongfaqiu/MVTImageryProvider/compare/v0.7.1...v0.7.2) (2022-09-15)


### Bug Fixes

* remove umd release ([c96fc3a](https://github.com/hongfaqiu/MVTImageryProvider/commit/c96fc3aef0e5831266eea5b69af8797d465c0f30))

## [0.7.1](https://github.com/hongfaqiu/MVTImageryProvider/compare/v0.7.0...v0.7.1) (2022-08-22)


### Bug Fixes

* fix pickFeatures when no features found ([5bad25c](https://github.com/hongfaqiu/MVTImageryProvider/commit/5bad25ce4430cd81fe577283b38326c0b4114deb))

# [0.7.0](https://github.com/hongfaqiu/MVTImageryProvider/compare/v0.6.2...v0.7.0) (2022-08-11)


### Features

* support EPSG4326 ([9536d0d](https://github.com/hongfaqiu/MVTImageryProvider/commit/9536d0d23178e3f2816f854995b4a2cbc81879a4))

## [0.6.2](https://github.com/hongfaqiu/MVTImageryProvider/compare/v0.6.1...v0.6.2) (2022-08-02)


### Bug Fixes

* some code syntax correction ([715fa74](https://github.com/hongfaqiu/MVTImageryProvider/commit/715fa74768d45172d5bea48016194419fe191f2e))

## [0.6.1](https://github.com/hongfaqiu/MVTImageryProvider/compare/v0.6.0...v0.6.1) (2022-07-28)


### Bug Fixes

* fix ci build ([67c29eb](https://github.com/hongfaqiu/MVTImageryProvider/commit/67c29eb490fc4be5d096fcbb44296164257c7fe0))

# [0.6.0](https://github.com/hongfaqiu/MVTImageryProvider/compare/v0.5.3...v0.6.0) (2022-07-27)


### Features

* add enablePickFeatures param ([33fd8c3](https://github.com/hongfaqiu/MVTImageryProvider/commit/33fd8c3431d9d8b1db7cd572e98130e1ac54f42e))


src/mapstyle.ts
// https://github.com/UW-Macrostrat/web/blob/main/src/map-interface/map-page/vector-style.ts

export const SETTINGS = {
  baseMapURL:
    "mapbox://styles/jczaplewski/cl3w3bdai001f14ob27ckmpxz?optimize=true",
  satelliteMapURL:
    "mapbox://styles/jczaplewski/cl51esfdm000e14mq51erype3?optimize=true",
  burwellTileDomain: 'https://tiles.macrostrat.org',
  mapboxAccessToken: 'pk.eyJ1IjoiamN6YXBsZXdza2kiLCJhIjoiY2szNXA5OWcxMDN2bzNtcnI1cWd1ZXJpYiJ9.Dd5GKlrPhg969y1ayY32cg',
};

export const mapStyle = {
  version: 8,
  sources: {
    burwell: {
      type: "vector",
      tiles: [`${SETTINGS.burwellTileDomain}/carto-slim/{z}/{x}/{y}.mvt`],
      tileSize: 512,
    },
  },
  layers: [
    {
      id: "burwell_fill",
      type: "fill",
      source: "burwell",
      "source-layer": "units",
      filter: ["!=", "color", ""],
      minzoom: 0,
      maxzoom: 16,
      paint: {
        "fill-color": {
          property: "color",
          type: "identity",
        },
        "fill-opacity": 0.4,

        // "fill-opacity": {
        //   "stops": [
        //     [0, 0.5],
        //     [12, 0.3]
        //   ]
        // }
      },
    },
    {
      id: "burwell_stroke",
      type: "line",
      source: "burwell",
      "source-layer": "units",
      filter: ["!=", "color", ""],
      minzoom: 0,
      maxzoom: 16,
      paint: {
        //  "line-color": "#777777",
        // "line-width": 0,
        "line-color": {
          property: "color",
          type: "identity",
        },
        "line-width": {
          stops: [
            [0, 0.15],
            [1, 0.15],
            [2, 0.15],
            [3, 0.15],
            [4, 0.2],
            [5, 0.4],
            [6, 0.05],
            [7, 0.1],
            [8, 0.4],
            [9, 0.5],
            [10, 0.35],
            [11, 0.4],
            [12, 1],
            [13, 1.25],
            [14, 1.5],
            [15, 1.75],
            [16, 2],
          ],
        },
        "line-opacity": {
          stops: [
            [0, 0],
            [4, 1],
          ],
        },
      },
    },
    // Hide water
    {
      id: "burwell_water_fill",
      type: "fill",
      source: "burwell",
      "source-layer": "units",
      filter: ["==", "color", ""],
      minzoom: 0,
      maxzoom: 16,
      paint: {
        "fill-opacity": 0,
      },
    },
    {
      id: "burwell_water_line",
      type: "line",
      source: "burwell",
      "source-layer": "units",
      filter: ["==", "color", ""],
      minzoom: 0,
      maxzoom: 16,
      paint: {
        "line-opacity": 0,
        "line-width": 1,
      },
    },
    {
      id: "faults",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: [
        "in",
        "type",
        "fault",
        "normal fault",
        "thrust fault",
        "strike-slip fault",
        "reverse fault",
        "growth fault",
        "fault zone",
        "zone",
      ],
      minzoom: 0,
      maxzoom: 16,
      paint: {
        "line-color": "#000000",
        "line-width": [
          "interpolate",
          ["linear"],
          ["zoom"],
          0,
          ["case", ["!=", ["get", "name"], ""], 0.6, 0.3],
          1,
          ["case", ["!=", ["get", "name"], ""], 0.6, 0.3],
          2,
          ["case", ["!=", ["get", "name"], ""], 0.6, 0.3],
          3,
          ["case", ["!=", ["get", "name"], ""], 0.6, 0.3],
          4,
          ["case", ["!=", ["get", "name"], ""], 1, 0.5],
          5,
          ["case", ["!=", ["get", "name"], ""], 1.2, 0.6],
          6,
          ["case", ["!=", ["get", "name"], ""], 0.9, 0.45],
          7,
          ["case", ["!=", ["get", "name"], ""], 0.8, 0.4],
          8,
          ["case", ["!=", ["get", "name"], ""], 1.4, 0.7],
          9,
          ["case", ["!=", ["get", "name"], ""], 1.6, 0.8],
          10,
          ["case", ["!=", ["get", "name"], ""], 1.4, 0.7],
          11,
          ["case", ["!=", ["get", "name"], ""], 2.2, 1.1],
          12,
          ["case", ["!=", ["get", "name"], ""], 2.6, 1.3],
          13,
          ["case", ["!=", ["get", "name"], ""], 3, 1.5],
          14,
          ["case", ["!=", ["get", "name"], ""], 3.2, 1.6],
          15,
          ["case", ["!=", ["get", "name"], ""], 3.5, 1.75],
          16,
          ["case", ["!=", ["get", "name"], ""], 4.4, 2.2],
        ],
        "line-opacity": 1,
      },
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
    },
    {
      id: "moraines",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "moraine"],
      minzoom: 12,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#3498DB",
        "line-dasharray": [1, 2],
        "line-width": {
          stops: [
            [10, 1],
            [11, 2],
            [12, 2],
            [13, 2.5],
            [14, 3],
            [15, 3],
          ],
        },
        "line-opacity": {
          stops: [
            [10, 0.2],
            [13, 1],
          ],
        },
      },
    },
    {
      id: "eskers",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "esker"],
      minzoom: 12,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#00FFFF",
        "line-dasharray": [1, 4],
        "line-width": {
          stops: [
            [10, 1],
            [11, 2],
            [12, 2],
            [13, 2.5],
            [14, 3],
            [15, 3],
          ],
        },
        "line-opacity": {
          stops: [
            [10, 0.2],
            [13, 1],
          ],
        },
      },
    },
    {
      id: "lineaments",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "lineament"],
      minzoom: 0,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#000000",
        "line-dasharray": [2, 2, 7, 2],
        "line-width": {
          stops: [
            [9, 1],
            [10, 1],
            [11, 2],
            [12, 2],
            [13, 2.5],
            [14, 3],
            [15, 3],
          ],
        },
        "line-opacity": 1,
      },
    },
    {
      id: "synclines",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "syncline"],
      minzoom: 0,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#F012BE",
        "line-width": {
          stops: [
            [0, 1],
            [7, 0.25],
            [8, 0.4],
            [9, 0.45],
            [10, 0.45],
            [11, 0.6],
            [12, 0.7],
            [13, 0.9],
            [14, 1.4],
            [15, 1.75],
            [16, 2.2],
          ],
        },
        "line-opacity": 1,
      },
    },
    {
      id: "monoclines",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "monocline"],
      minzoom: 0,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#F012BE",
        "line-width": {
          stops: [
            [0, 1],
            [7, 0.25],
            [8, 0.4],
            [9, 0.45],
            [10, 0.45],
            [11, 0.6],
            [12, 0.7],
            [13, 0.9],
            [14, 1.4],
            [15, 1.75],
            [16, 2.2],
          ],
        },
        "line-opacity": 1,
      },
    },
    {
      id: "folds",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "fold"],
      minzoom: 0,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#F012BE",
        "line-width": {
          stops: [
            [0, 1],
            [7, 0.25],
            [8, 0.4],
            [9, 0.45],
            [10, 0.45],
            [11, 0.6],
            [12, 0.7],
            [13, 0.9],
            [14, 1.4],
            [15, 1.75],
            [16, 2.2],
          ],
        },
        "line-opacity": 1,
      },
    },
    {
      id: "dikes",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "dike"],
      minzoom: 6,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#FF4136",
        "line-width": {
          stops: [
            [0, 1],
            [7, 0.25],
            [8, 0.4],
            [9, 0.45],
            [10, 0.45],
            [11, 0.6],
            [12, 0.7],
            [13, 0.9],
            [14, 1.4],
            [15, 1.75],
            [16, 2.2],
          ],
        },
        "line-opacity": {
          stops: [
            [6, 0.2],
            [10, 1],
          ],
        },
      },
    },
    {
      id: "anticlines",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "anticline"],
      minzoom: 0,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#F012BE",
        "line-width": {
          stops: [
            [0, 1],
            [7, 0.25],
            [8, 0.4],
            [9, 0.45],
            [10, 0.45],
            [11, 0.6],
            [12, 0.7],
            [13, 0.9],
            [14, 1.4],
            [15, 1.75],
            [16, 2.2],
          ],
        },
        "line-opacity": 1,
      },
    },
    {
      id: "flows",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "flow"],
      minzoom: 0,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#FF4136",
        "line-width": {
          stops: [
            [0, 1],
            [7, 0.25],
            [8, 0.4],
            [9, 0.45],
            [10, 0.45],
            [11, 0.6],
            [12, 0.7],
            [13, 0.9],
            [14, 1.4],
            [15, 1.75],
            [16, 2.2],
          ],
        },
        "line-opacity": 1,
      },
    },
    {
      id: "sills",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "sill"],
      minzoom: 0,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#FF4136",
        "line-width": {
          stops: [
            [0, 1],
            [7, 0.25],
            [8, 0.4],
            [9, 0.45],
            [10, 0.45],
            [11, 0.6],
            [12, 0.7],
            [13, 0.9],
            [14, 1.4],
            [15, 1.75],
            [16, 2.2],
          ],
        },
        "line-opacity": 1,
      },
    },
    {
      id: "veins",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["==", "type", "vein"],
      minzoom: 0,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#FF4136",
        "line-width": {
          stops: [
            [0, 1],
            [7, 0.25],
            [8, 0.4],
            [9, 0.45],
            [10, 0.45],
            [11, 0.6],
            [12, 0.7],
            [13, 0.9],
            [14, 1.4],
            [15, 1.75],
            [16, 2.2],
          ],
        },
        "line-opacity": {
          stops: [
            [6, 0.2],
            [10, 1],
          ],
        },
      },
    },
    {
      id: "marker_beds",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["in", "type", "marker bed", "bed"],
      minzoom: 12,
      maxzoom: 16,
      layout: {
        "line-join": "round",
        "line-cap": "round",
      },
      paint: {
        "line-color": "#333333",
        "line-width": {
          stops: [
            [10, 0.8],
            [11, 0.8],
            [12, 0.9],
            [13, 0.9],
            [14, 1.4],
            [15, 1.75],
            [16, 2.2],
          ],
        },
        "line-opacity": 1,
      },
    },
    {
      id: "craters",
      type: "line",
      source: "burwell",
      "source-layer": "lines",
      filter: ["in", "type", "crater", "impact structure"],
      minzoom: 10,
      maxzoom: 16,
      paint: {
        "line-dasharray": [6, 6],
        "line-color": "#000000",
        "line-width": {
          stops: [
            [10, 0.6],
            [11, 0.6],
            [12, 0.72],
            [13, 0.72],
            [14, 1],
            [15, 1.3],
            [16, 1.8],
          ],
        },
        "line-opacity": 1,
      },
    },
  ],
};
